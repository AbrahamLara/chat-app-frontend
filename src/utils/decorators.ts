import { createDecorator } from 'vue-class-component';
import { DispatchOptions } from 'vuex';

/**
 * A custom decorator to turn properties into methods that can dispatch actions.
 * ***
 * _Note:_ The name provided is the action that will be dispatched. If none is provided, then the property name this
 * decorator is attached to will be used as the action name. Meaning the name of the property should be exactly the
 * name of the action intended to be dispatched.
 * ***
 * Example:
 * ```
 * @Component
 * class Component extends Vue {
 *   @Action('actionName') callAction!;
 *   // or
 *   @Action() actionName!;
 * }
 * ```
 *
 * @param actionName The action to dispatch.
 * @param namespace The name.
 * @return Returns a function that accepts a payload and options.
 */
export function Action(actionName?: string, namespace?: string) {
  return createDecorator((options, property) => {
    if (!options.methods) {
      options.methods = {};
    }
    let action: string;
    if (actionName && namespace) {
      action = `${namespace}/${actionName}`;
    } else if (actionName) {
      action = actionName;
    } else {
      action = property;
    }
    // The property name will be used as the method name since the property will be used to dispatch the action.
    options.methods[property] = function(
      payload,
      options?: DispatchOptions | undefined
    ): Promise<any> {
      return this.$store.dispatch(action, payload, options);
    };
  });
}
