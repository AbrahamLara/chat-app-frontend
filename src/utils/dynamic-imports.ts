// This file contains dynamic imports of components to reduce bundle size when building project with webpack.

// Components
const getAbout = () =>
  import(/* webpackChunkName: "About" */ '@/components/About.vue');

const getAppBar = () =>
  import(/* webpackChunkName: "AppBar" */ '@/components/AppBar.vue');

const getFormCard = () =>
  import(/* webpackChunkName: "FormCard" */ '@/components/FormCard.vue');

const getRegexRulesTester = () =>
  import(
    /* webpackChunkName: "RegexRulesTester" */ '@/components/RegexRulesTester.vue'
  );

const getSettings = () =>
  import(/* webpackChunkName: "Settings" */ '@/components/Settings.vue');

const getSignIn = () =>
  import(/* webpackChunkName: "SignIn" */ '@/components/SignIn.vue');

const getSignUp = () =>
  import(/* webpackChunkName: "SignUp" */ '@/components/SignUp.vue');

// Pages
const getAuthPage = () =>
  import(/* webpackChunkName: "AuthPage"  */ '@/views/AuthPage.vue');

const getNotFound = () =>
  import(/* webpackChunkName: "NotFoundPage" */ '@/views/NotFoundPage.vue');

const getSplash = () =>
  import(/* webpackChunkName: "SplashPage" */ '@/views/SplashPage.vue');

const getChatPage = () =>
  import(/* webpackChunkName: "ChatPage" */ '@/views/ChatPage.vue');

export {
  getAbout,
  getAppBar,
  getAuthPage,
  getFormCard,
  getNotFound,
  getRegexRulesTester,
  getSettings,
  getSignIn,
  getSignUp,
  getSplash,
  getChatPage,
};
