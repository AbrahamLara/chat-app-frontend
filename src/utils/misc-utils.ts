/**
 * An user item that can be used for contexts such as displaying a user search or chat result.
 */
import Vue from 'vue';

export interface UserItem {
  /**
   * The name of the user.
   */
  name: string;

  /**
   * The timestamp to determine the duration of time relative to today.
   */
  timeStamp?: string | number;

  /**
   * A user message.
   */
  message?: string;
}

/**
 * Returns a cookie stored in the page with the given name.
 */
export function getBrowserCookie(name: string) {
  return Vue.$cookies.get(name);
}
