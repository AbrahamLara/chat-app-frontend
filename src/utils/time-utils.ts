import TimeAgo from 'javascript-time-ago';

const timeAgo = new TimeAgo('en-US');

/**
 * Returns a twitter "time ago" label using the provided date/time value. This helps display how much time has elapsed
 * from the given date/time string.
 *
 * @param value date/time string.
 */
export function getTwitterTimeAgoLabel(value: string | number) {
  return timeAgo.format(new Date(value), 'twitter');
}
