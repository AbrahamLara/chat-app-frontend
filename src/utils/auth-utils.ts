import { UserState } from '@/store/store-states';
import { DEFAULT_ALERTS_STATE } from '@/store/modules/alerts-module';
import { cloneObject } from '@/utils/misc-utils';

/**
 * Login form fields definition.
 */
export interface LoginFormFields {
  email: string;
  password: string;
}
export type LoginField = keyof LoginFormFields;

/**
 * Registered form fields definition.
 */
export interface RegisterFormFields extends LoginFormFields {
  name: string;
  /**
   * The confirmation password that should match the value of the password field.
   */
  confPassword: string;
}
export type RegisterFormField = keyof RegisterFormFields;

/**
 * The representation of a regex rule that has passed or not.
 */
export interface RegexRule {
  regex: RegExp;

  /**
   * The rule the represents regex being tested.
   */
  rule: string;

  /**
   * Determines if the regex test passed or not.
   */
  passed?: boolean;
}

export interface UserProfile {
  /**
   * The id of the authenticated user.
   */
  id: string;

  /**
   * The name of the authenticated user.
   */
  name: string;

  /**
   * The email of the authenticated user.
   */
  email: string;
}

/**
 * Password regex also used in the backend.
 * [PASSWORD_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L22-L41)
 */
export const PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*\-_]).{8,}$/;

/**
 * Email regex also used in the backend.
 * [EMAIL_REGEX](https://github.com/AbrahamLara/chat-app-backend/blob/6757dd186701f4581f931c957d8bccd7840d4d3d/utils/misc.ts#L43-L48)
 */
export const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

// The rules for validating the email text field.
export const EMAIL_FORM_RULES = [
  (email: string) => Boolean(email) || 'Email is required',
  (email: string) => EMAIL_REGEX.test(email) || 'Email must be valid.',
];

// The rules for validating the password text field.
export const PASSWORD_FORM_RULES = [
  (password: string) => Boolean(password) || 'Password is required',
  (password: string) =>
    PASSWORD_REGEX.test(password) || 'Password must be valid.',
];

// The rules for validating the password regex.
export const PASSWORD_REGEX_RULES: RegexRule[] = [
  { rule: '8 characters minimum', regex: /.{8,}/ },
  { rule: '1 uppercase character', regex: /[A-Z]/ },
  { rule: '1 lowercase character', regex: /[a-z]/ },
  { rule: '1 number', regex: /[0-9]/ },
  {
    rule: '1 special character: !, @, #, $, %, ^, &, *, -, _',
    regex: /[!@#$%^&*\-_]/,
  },
];

export const DEFAULT_USER_PROFILE = {
  id: '',
  name: '',
  email: '',
};

export const DEFAULT_USER_STATE: UserState = {
  ...cloneObject(DEFAULT_ALERTS_STATE),
  isAuthenticated: false,
  user: DEFAULT_USER_PROFILE,
};

export const DEFAULT_LOGIN_FORM_FIELDS: LoginFormFields = {
  email: '',
  password: '',
};

export const DEFAULT_REGISTER_FORM_FIELDS: RegisterFormFields = {
  name: '',
  email: '',
  password: '',
  confPassword: '',
};

export const INVALID_TOKEN_MESSAGE = 'The provided token is invalid';

/**
 * Determines if the provided object is a message about an invalid token.
 */
export function isInvalidTokenMessage(data: Record<string, string>) {
  return data.message === 'The provided token is invalid';
}
