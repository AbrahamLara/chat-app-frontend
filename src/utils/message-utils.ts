import { MessagesState } from '@/store/store-states';
import { DEFAULT_ALERTS_STATE } from '@/store/modules/alerts-module';
import { UserProfile } from '@/utils/auth-utils';

export const DEFAULT_MESSAGES_STATE: MessagesState = {
  ...DEFAULT_ALERTS_STATE,
  map: new Map(),
};

export const FAILED_TO_FETCH_MESSAGES_ERROR =
  'An error occurred fetching chat messages';

export const FAILED_TO_SEND_MESSAGE_ERROR =
  'An error occurred sending chat messages';

export type MessageAuthor = Omit<UserProfile, 'email'>;

export interface MessageItem {
  /**
   * The id of the chat message.
   */
  id: string;

  /**
   * The chat message text sent by the author.
   */
  text: string;

  /**
   * The author of the chat message.
   */
  author: MessageAuthor;

  /**
   * The timestamp of when the chat message was created.
   */
  createdAt: string;
}

export interface UserTypingMessageData {
  /**
   * The id of the chat to broadcast to that a member is typing a message.
   */
  chatID: string;

  /**
   * The name of the user typing a message to the chat.
   */
  user: MessageAuthor;

  /**
   * Determines if the user is typing.
   */
  isTyping: boolean;
}

export interface SendChatMessageData {
  /**
   * The id of the chat to broadcast the newly created message to.
   */
  chatID: string;

  /**
   * The object of the message to send to other user's in the chat.
   */
  message: MessageItem;
}
