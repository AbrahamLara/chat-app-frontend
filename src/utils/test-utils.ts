import VueRouter, { RouteConfig } from 'vue-router';
import Vuex from 'vuex';
import { AppState } from '@/store/store-states';
import appMutations from '@/store/mutations';
import appActions from '@/store/actions';
import appModules from '@/store/modules';
import { DEFAULT_ROOT_STATE } from '@/store';
import { FormAlertMessage } from '@/utils/alerts-utils';
import { createLocalVue } from '@vue/test-utils';
import VueCookies from 'vue-cookies';
import Vue from 'vue';
import Vuetify from 'vuetify';

/**
 * Describes the available options for setting up a local vue instance for a test file.
 */
interface LocalVueSetupOptions {
  /**
   * Indicates that vuex plugin should be used.
   */
  useVuex?: boolean;

  /**
   * Indicates that vuetify plugin should be used.
   */
  useVuetify?: boolean;

  /**
   * Indicates that vue cookies plugin should be used.
   */
  useVueCookies?: boolean;

  /**
   * Indicates that vue router plugin should be used.
   */
  useVueRouter?: boolean;
}

/**
 * Resolves a vuetify warning of not being able to locate a data-app attribute on an element because vuetify
 * components should be wrapped by a v-app component.
 *
 * ```
 * Ex: [Vuetify] Unable to locate target [data-app] in "v-menu"
 * ```
 */
export function resolveVuetifyAppDataWarning() {
  const app = document.createElement('div');
  app.setAttribute('data-app', 'true');
  document.body.append(app);
}

/**
 * Create a Vuex store instance for testing.
 *
 * TODO: Reuse to create store in main.ts.
 *
 * @param vx The Vuex value
 * @param customState A state object that modifies the default state
 */
export function createStore(
  vx: typeof Vuex,
  customState: Partial<AppState> = {}
) {
  const mutations = appMutations;
  const actions = appActions;
  const modules = appModules;

  return new vx.Store({
    state: { ...DEFAULT_ROOT_STATE, ...customState },
    actions,
    mutations,
    modules,
  });
}

/**
 * Creates a VueRouter instance for testing.
 *
 * TODO: Reuse to create router in main.ts.
 *
 * @param vr The VueRouter value
 * @param routes The route config. If not provided, then a default config is provided.
 */
export function createRouter(
  vr: typeof VueRouter,
  routes: RouteConfig[] = [{ path: '/' }]
) {
  return new vr({ routes, mode: 'history' });
}

/**
 * Creates a form error message for testing.
 */
export function createFormAlertMessage(
  field: string,
  message: string
): FormAlertMessage {
  return { field, message };
}

/**
 * Creates a local vue instance with options to use any vue plugins.
 */
export function createLocalVueInstance(options: LocalVueSetupOptions = {}) {
  const { useVueCookies, useVuex, useVuetify, useVueRouter } = options;
  const localVue = createLocalVue();

  if (useVueCookies) {
    Vue.use(VueCookies);
  }
  if (useVuex) {
    localVue.use(Vuex);
  }
  if (useVuetify) {
    Vue.use(Vuetify);
  }
  if (useVueRouter) {
    localVue.use(VueRouter);
  }

  return localVue;
}
