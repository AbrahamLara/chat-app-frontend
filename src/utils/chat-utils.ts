import { ChatsState } from '@/store/store-states';
import { DEFAULT_ALERTS_STATE } from '@/store/modules/alerts-module';
import { MessageAuthor, MessageItem } from '@/utils/message-utils';
import { cloneObject } from '@/utils/misc-utils';

export const DEFAULT_CHATS_STATE: ChatsState = {
  ...cloneObject(DEFAULT_ALERTS_STATE),
  chats: [],
  userTypingMap: new Map<string, MessageAuthor[]>(),
  currentChatID: '',
};

export const CREATE_CHAT_SUCCESS_MESSAGE = 'Successfully created chat';

export interface CreateChatFormFields {
  /**
   * The name to give the new chat.
   */
  chatName: string;

  /**
   * The first message to send to the chat.
   */
  message: string;

  /**
   * An array of user ids from members to add to the group chat.
   */
  userIDs: string[];
}

export interface UserChatItem {
  /**
   * The id of the user chat.
   */
  id: string;

  /**
   * The name of the chat.
   */
  name: string;

  /**
   * The object representing the chat message sent to the chat.
   */
  message: Omit<MessageItem, 'author'> & { author: string };
}

export interface ChatCreatedData {
  chat: UserChatItem;

  /**
   * A list of user ids from members part of the newly created chat.
   */
  memberIDs: string[];
}
