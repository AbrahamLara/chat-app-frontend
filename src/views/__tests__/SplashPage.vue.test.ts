import Vuetify from 'vuetify';
import Splash from '@/views/SplashPage.vue';
import { mount, Wrapper } from '@vue/test-utils';
import { createLocalVueInstance } from '@/utils/test-utils';

const localVue = createLocalVueInstance({ useVuetify: true });

describe('SplashPage.vue', () => {
  let wrapper: Wrapper<any>;

  beforeEach(() => {
    const vuetify = new Vuetify();

    wrapper = mount(Splash, { localVue, vuetify });
  });

  it('renders splash view correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('contains about view', () => {
    expect(wrapper.find('.about-container').exists()).toBeTruthy();
  });
});
