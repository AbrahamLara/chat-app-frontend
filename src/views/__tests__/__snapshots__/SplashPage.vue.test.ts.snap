// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplashPage.vue renders splash view correctly 1`] = `
<div
  class="splash-view d-flex flex-column-reverse flex-md-row"
>
  <div
    class="d-lg-flex align-center justify-center col-md-8 col-lg-6 col"
  >
    <div
      class="about-container pa-sai pt-5 pt-lg-0"
    >
      <div
        class="headline"
      >
        About
      </div>
       
      <div
        class="body-1 pt-3"
      >
        
    ChatApp is a mock WhatsApp application which this project was also
    inspired by. The idea behind this project was to recreate the basic
    features and functionalities that exist within WhatsApp.
  
      </div>
       
      <div
        class="body-1 pt-3"
      >
        
    I mainly started this project in order to demonstrate and apply what I've
    learned working with the tools and technologies used to create this
    project.
  
      </div>
       
      <div
        class="headline mt-9"
      >
        Description
      </div>
       
      <div
        class="body-1 pt-3"
      >
        
    Users are able to Signup and Login to begin chatting with other users on
    the platform. Users will be able to tell when other users are typing in
    real-time.
  
      </div>
       
      <div
        class="headline mt-9"
      >
        Technologies
      </div>
       
      <div
        class="ml-3"
      >
        <div
          class="heading pt-3"
        >
          Frontend
        </div>
         
        <ul
          class="ml-2"
        >
          <li>
            TypeScript
          </li>
           
          <li>
            Vue
          </li>
           
          <li>
            Sass
          </li>
           
          <li>
            Webpack
          </li>
           
          <li>
            Jest
          </li>
        </ul>
      </div>
       
      <div
        class="ml-3"
      >
        <div
          class="heading pt-3"
        >
          Backend
        </div>
         
        <ul
          class="ml-2"
        >
          <li>
            Typescript
          </li>
           
          <li>
            Express.js
          </li>
           
          <li>
            Docker
          </li>
           
          <li>
            Sequelize
          </li>
           
          <li>
            Jest
          </li>
        </ul>
      </div>
       
      <div
        class="heading pt-3 ml-3"
      >
        Database â€” PostgreSQL
      </div>
    </div>
  </div>
   
  <div
    class="splash-view_welcome-panel d-flex flex-column align-center justify-center col-md-4 col-lg-6 col"
  >
    <div
      class="splash-card-thing text-h4 text-lg-h3 pa-3 pt-1 pb-1 pa-lg-5 mr-16 v-card v-card--flat v-sheet theme--light rounded-0"
    >
      
      Welcome
    
    </div>
     
    <div
      class="splash-card-thing text-h4 text-lg-h3 pa-3 pt-1 pb-1 pa-lg-5 mt-4 v-card v-card--flat v-sheet theme--light rounded-0"
    >
      
      to
    
    </div>
     
    <div
      class="splash-card-thing text-h4 text-lg-h3 pa-3 pt-1 pb-1 pa-lg-5 mt-4 ml-16 v-card v-card--flat v-sheet theme--light rounded-0"
    >
      
      ChatApp!
    
    </div>
  </div>
</div>
`;
